gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,568 @@
+          if (bytesRead < 0) {
+            logger.atInfo().log(
+                "Somehow read %d bytes trying to skip %d bytes to seek to position %d, size: %d",
+                bytesRead, seekDistance, currentPosition, objectSize);
+            closeContentChannel();
+          } else {
+            seekDistance -= bytesRead;
+            contentChannelCurrentPosition += bytesRead;
+          }
+        } catch (Exception e) {
+          logger.atInfo().withCause(e).log(
+              "Got an IO exception on contentChannel.read(), a lazy-seek will be pending for '%s'",
+              resourceId);
+          closeContentChannel();
+        }
+      }
+      checkState(
+          byteChannel == null || contentChannelCurrentPosition == currentPosition,
+          "contentChannelPosition (%s) should be equal to currentPosition (%s)"
+              + " after successful in-place skip",
+          contentChannelCurrentPosition,
+          currentPosition);
+    }
+
+    private void performPendingSeeks() {
+
+      // Return quickly if there is no pending seek operation, i.e. position didn't change.
+      if (currentPosition == contentChannelCurrentPosition && byteChannel != null) {
+        return;
+      }
+
+      logger.atFiner().log(
+          "Performing lazySeek from %s to %s position '%s'",
+          contentChannelCurrentPosition, currentPosition, resourceId);
+
+      if (isInRangeSeek()) {
+        skipInPlace();
+      } else {
+        if (isRandomAccessPattern()) {
+          setRandomAccess();
+        }
+        // close existing contentChannel as requested bytes can't be served from current
+        // contentChannel;
+        closeContentChannel();
+      }
+    }
+
+    private boolean isRandomAccessPattern() {
+      if (!shouldDetectRandomAccess()) {
+        return false;
+      }
+      if (currentPosition < contentChannelCurrentPosition) {
+        logger.atFine().log(
+            "Detected backward read from %s to %s position, switching to random IO for '%s'",
+            contentChannelCurrentPosition, currentPosition, resourceId);
+        return true;
+      }
+      if (contentChannelCurrentPosition >= 0
+          && contentChannelCurrentPosition + readOptions.getInplaceSeekLimit() < currentPosition) {
+        logger.atFine().log(
+            "Detected forward read from %s to %s position over %s threshold,"
+                + " switching to random IO for '%s'",
+            contentChannelCurrentPosition,
+            currentPosition,
+            readOptions.getInplaceSeekLimit(),
+            resourceId);
+        return true;
+      }
+      return false;
+    }
+
+    private ReadableByteChannel getStorageReadChannel(long seek, long limit) throws IOException {
+      ReadChannel readChannel = storage.reader(blobId, generateReadOptions(blobId));
+      try {
+        readChannel.seek(seek);
+        readChannel.limit(limit);
```

Comment:
why not assert that limit is not less than zero?

Reply:
Library does that for us and will throw. We have unit test covering this scenario.